### PR TITLE
DAOS-16722 client: to intercept PMPI_Init() in libpil4dfs

### PR DESCRIPTION
**Pull request by wiliamhuang**
Intercept PMPI_Init() to avoid calling daos_init() if MPI_Init() is intercepted by other library (like darshan and mpip).

Features: pil4dfs

Required-githooks: true
Skipped-githooks: codespell

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by mchaarawi — APPROVED** [review]

**Comment by wiliamhuang** [comment]
All CI tests finished without issues.

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 1]
```diff
@@ -1057,6 +1058,22 @@ MPI_Init(int *argc, char ***argv)
 	return rc;
 }
 
+int
+PMPI_Init(int *argc, char ***argv)
+{
+	int rc;
+
+	if (next_pmpi_init == NULL) {
+		next_pmpi_init = dlsym(RTLD_NEXT, "PMPI_Init");
```
Comment (knard-intel) [inline src/client/dfuse/pil4dfs/int_dfs.c:1067]: For my understanding, we are expecting to get the reference on the original PMPI_Init function.
However, I do understand how we are sure of which function reference we will get, if there is several LD_PRELOAD on this last one.  For example, if Darshan is redefining it, could it be possible that we get the reference on this last one.

**Review by daltonbohning — APPROVED** [review]

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 1 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1067]: The issue of deadlock is caused by calling daos_init() inside MPI runtime library's MPI_Init()/PMPI_Init(). We could avoid this issue as long as the PMPI_Init() from libpil4dfs is executed before the PMPI_Init() implemented in MPI runtime library.  It does not not matter the PMPI_Init() from libpil4dfs or the PMPI_Init() from darshan is executed first.

Comment (knard-intel) [reply src/client/dfuse/pil4dfs/int_dfs.c:1067]: OK, thanks for the explanation.

**Review by knard-intel — APPROVED** [review]

**Comment by mchaarawi** [comment]
let's hold off merging for now until i test on aurora since this is not very time critical at the moment.
i have access to darshan there so should be quick once i get an ECB.

**Comment by wiliamhuang** [comment]
> let's hold off merging for now until i test on aurora since this is not extremelly time critical anymore. i have access to darshan there so should be quick once i get an ECB.

Thank you very much!

**Pull request merged**